### ISSUE TITLE
the loss compute problems in rnn

### ISSUE THREAD

**Issue by tttwwy** [fchollet/keras#765]
I want to use rnn in  language model, is there any different between the two models below when computing the loss? In my opinion, if my input is [1,2,3,4],the first model will compute the loss of  input [1] ,[1,2] ,[1,2,3],[1,2,3,4] and sum them up ,the second model will only compute the loss of input [ 1,2,3,4] , am I right?
```py
model.add(GRU(embedding_dim, hidden_dim, activation="relu",return_sequences=True))
model.add(TimeDistributedDense(hidden_dim, input_dim))
model.add(Activation('softmax'))
model.compile(loss='categorical_crossentropy', optimizer='rmsprop')
```

```py
model.add(GRU(embedding_dim, hidden_dim, activation="relu",return_sequences=False))
model.add(Dense(hidden_dim, input_dim))
model.add(Activation('softmax'))
model.compile(loss='categorical_crossentropy', optimizer='rmsprop')
```

**Comment by tttwwy** [comment]
I want to use rnn in  language model, is there any different between the two models below when computing the loss? In my opinion, if my input is [1,2,3,4],the first model will compute the loss of  input [1] ,[1,2] ,[1,2,3],[1,2,3,4] and sum them up ,the second model will only compute the loss of input [ 1,2,3,4] , am I right?
```py
model.add(GRU(embedding_dim, hidden_dim, activation="relu",return_sequences=True))
model.add(TimeDistributedDense(hidden_dim, input_dim))
model.add(Activation('softmax'))
model.compile(loss='categorical_crossentropy', optimizer='rmsprop')
```

```py
model.add(GRU(embedding_dim, hidden_dim, activation="relu",return_sequences=False))
model.add(Dense(hidden_dim, input_dim))
model.add(Activation('softmax'))
model.compile(loss='categorical_crossentropy', optimizer='rmsprop')
```

**Comment by EderSantana** [comment]
did you see this example https://github.com/fchollet/keras/blob/master/examples/lstm_text_generation.py ??

`TimeDistributedDense` is the safer way to go since it is a `MaskedLayer` which means, it passes the masking forward in case of different length input output.

**Comment by tttwwy** [comment]
can dense layer support the mask？


来自 魅族 MX5

-------- 原始邮件 --------
发件人：Eder Santana <notifications@github.com>
时间：周六 10月3日 00:00
收件人：fchollet/keras <keras@noreply.github.com>
抄送：xiaose <tttwwy@126.com>
主题：Re: [keras] the loss compute problems in rnn (#765)

**Comment by fchollet** [comment]
@tttwwy you can using `Masking` layers anywhere in a network.
